System/Metrics/Prometheus/Validation.hs

```diff
@@ -0,0 +1,108 @@
+        Nothing -> False
+        Just (headChar, _tail) ->
+            isInitialNameChar headChar && T.all isNameChar name
+
+isNameChar :: Char -> Bool
+isNameChar c = isInitialNameChar c || isDigit c
+
+isInitialNameChar :: Char -> Bool
+isInitialNameChar c =
+  'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || c == '_'
+
+--------------------------------------------------------------------------------
+
+-- | Test whether a string is valid Prometheus help text.
+--
+-- According to <https://prometheus.io/docs/instrumenting/exposition_formats/>,
+-- help text may contain any sequence of UTF-8 characters, but the backslash
+-- (@\\@) and line feed (@\\n@) characters have to be escaped as @\\\\@ and
+-- @\\n@,
+-- respectively.
+isValidHelpText :: T.Text -> Bool
+isValidHelpText =
+  -- Using a strict left fold because we almost always expect a valid string,
+  -- for which verification requires traversing the entire string.
+  checkResult . T.foldl' f initialCheckState
+  where
+    f :: CheckState -> Char -> CheckState
+    f CheckState{haveFoundError, isLastCharBackslash} char =
+      case char of
+        '\n' -> CheckState True NotBackslash
+        '\\' ->
+          if isLastCharBackslash == Backslash
+            then CheckState haveFoundError NotBackslash
+            else CheckState haveFoundError Backslash
+        'n' -> CheckState haveFoundError NotBackslash
+        _ -> CheckState
+              (isLastCharBackslash == Backslash || haveFoundError)
+              NotBackslash
+
+-- | Test whether a string is a valid Prometheus label value.
+--
+-- According to <https://prometheus.io/docs/instrumenting/exposition_formats/>,
+-- label values can be any sequence of UTF-8 characters, but the backslash
+-- (@\\@), double-quote (@"@), and line feed (@\\n@) characters have to be
+-- escaped as @\\\\@, @\\"@, and @\\n@, respectively.
+isValidLabelValue :: T.Text -> Bool
+isValidLabelValue =
+  -- Using a strict left fold because we almost always expect a valid string,
+  -- for which verification requires traversing the entire string.
+  checkResult . T.foldl' f initialCheckState
+  where
+    f :: CheckState -> Char -> CheckState
+    f CheckState{haveFoundError, isLastCharBackslash} char =
+      case char of
+        '\n' -> CheckState True NotBackslash
+        '\\' ->
+          if isLastCharBackslash == Backslash
+            then CheckState haveFoundError NotBackslash
+            else CheckState haveFoundError Backslash
+        'n' -> CheckState haveFoundError NotBackslash
+        '\"' ->
+          -- Extra case relative to `isValidHelpText`
+          if isLastCharBackslash == Backslash
+            then CheckState haveFoundError NotBackslash
+            else CheckState True NotBackslash
+        _ -> CheckState
+              (isLastCharBackslash == Backslash || haveFoundError)
+              NotBackslash
+
+data CheckState = CheckState
+  { haveFoundError :: !Bool
+  , isLastCharBackslash :: !IsBackslash
```

Comment:
```suggestion
  , isPrevCharBackslash :: !IsBackslash
```

would be a little clearer imo, but doesn't really matter

Reply:
This is much better, thanks!